File: python/repuwave_sdk/service.py
"""
Repuwave Python SDK — Service Client.

Client for services to verify agent signatures and report outcomes.
"""

from __future__ import annotations

import hashlib
import time

import httpx


class RepuwaveService:
    """
    Client for services receiving requests from Repuwave-authenticated agents.
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.repuwave.io/v1",
        **httpx_kwargs,
    ) -> None:
        self._api_key = api_key
        self._client = httpx.Client(
            base_url=base_url,
            headers={"X-Service-API-Key": api_key},
            **httpx_kwargs,
        )

    def verify(
        self,
        uaid: str,
        signature: str,
        timestamp: str | float,
    ) -> dict:
        """
        Verify an agent's signature and retrieve their current trust score.
        """
        response = self._client.get(
            f"/verify/{uaid}/",
            headers={
                "X-Repuwave-Signature": signature,
                "X-Repuwave-Timestamp": str(timestamp),
            },
        )
        if response.status_code == 404:
            return {"verified": False, "score": 0, "trust_level": "UNKNOWN", "uaid": uaid}

        response.raise_for_status()
        return response.json()

    def report(
        self,
        uaid: str,
        event_type: str,
        weight: float,
        signature: str,
        agent_timestamp: float,
        body: bytes | None = None,
        body_hash: str = "",
    ) -> dict:
        """
        Report an interaction outcome for an agent.

        Every report must carry proof of interaction — the Ed25519 signature
        the agent sent with its original request (X-Repuwave-Signature),
        together with the signed timestamp (X-Repuwave-Timestamp) and the
        SHA-256 hash of the raw request body the agent signed.

        Args:
            uaid: The agent's UAID.
            event_type: One of TXN_SUCCESS, TXN_HV_SUCCESS, TXN_NEUTRAL,
                TXN_FAIL_MINOR, TXN_FAIL_MAJOR, TXN_MALICIOUS, TXN_CHARGEBACK.
            weight: Contextual weight of the interaction (0.1–10.0).
            signature: The agent's signature from the original request.
            agent_timestamp: The Unix timestamp the agent signed.
            body: Raw request body bytes the agent signed (hashed for you).
            body_hash: Precomputed SHA-256 hex of the body (alternative to
                `body`; leave both empty for bodyless requests).

        Returns:
            The API response, e.g. {"accepted": True, "event_id": "..."}.
        """
        if body and not body_hash:
            body_hash = hashlib.sha256(body).hexdigest()

        response = self._client.post(
            "/report/",
            json={
                "uaid": uaid,
                "event_type": event_type,
                "weight": weight,
                "signature": signature,
                "agent_timestamp": agent_timestamp,
                "body_hash": body_hash,
                "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            },
        )
        response.raise_for_status()
        return response.json()

    def close(self) -> None:
        self._client.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

File: node/src/service.ts
/**
 * Repuwave Node SDK — Service Client.
 *
 * For services that receive requests from Repuwave-authenticated agents:
 * verify an agent's trust score and report interaction outcomes back to
 * Repuwave. Authenticates with the service API key via the X-Service-API-Key
 * header. Uses the global fetch (Node 18+) — no dependencies.
 */

import { createHash } from "crypto";
import { ApiError } from "./errors";

export interface RepuwaveServiceOptions {
  /** The service's API key (X-Service-API-Key). */
  apiKey: string;
  baseUrl?: string;
}

export interface VerifyParams {
  uaid: string;
  /**
   * The agent's signature from the incoming request.
   *
   * Required. The server checks it against the agent's registered public key
   * and refuses without it -- a UAID alone proves nothing, because UAIDs are
   * public, listed in the key directory. This was typed optional and described
   * as "optional context" while the endpoint ignored it entirely.
   */
  signature: string;
  /** The timestamp the agent signed. Valid 15s either side of server time. */
  timestamp: string | number;
  /**
   * sha256 of the body the agent sent you, if its request had one. The agent
   * signed that hash and only you saw the body, so without it the signature
   * cannot be reconstructed.
   */
  bodyHash?: string;
}

export type EventType =
  | "TXN_SUCCESS"
  | "TXN_HV_SUCCESS"
  | "TXN_NEUTRAL"
  | "TXN_FAIL_MINOR"
  | "TXN_FAIL_MAJOR"
  | "TXN_MALICIOUS"
  | "TXN_CHARGEBACK";

export interface ReportParams {
  uaid: string;
  eventType: EventType;
  weight: number;
  /** The agent's signature from its original request (proof of interaction). */
  signature: string;
  /** The Unix timestamp the agent signed (X-Repuwave-Timestamp). */
  agentTimestamp: number;
  /** Raw request body the agent signed (hashed for you), or a precomputed hash. */
  body?: string;
  bodyHash?: string;
}

export class RepuwaveService {
  private readonly apiKey: string;
  private readonly baseUrl: string;

  constructor(options: RepuwaveServiceOptions) {
    this.apiKey = options.apiKey;
    this.baseUrl = (options.baseUrl || "https://repuwave.fasolink.app/v1").replace(/\/$/, "");
  }

  private headers(extra: Record<string, string> = {}): Record<string, string> {
    return { "X-Service-API-Key": this.apiKey, ...extra };
  }

  private async parseError(response: Response): Promise<never> {
    let code = "unknown_error";
    let message = response.statusText;
    let body: unknown = null;
    try {
      body = await response.json();
      const err = (body as { error?: { code?: string; message?: string } }).error;
      if (err) {
        code = err.code ?? code;
        message = err.message ?? message;
      }
    } catch {
      /* non-JSON error body */
    }
    throw new ApiError(message, response.status, code, body);
  }

  /**
   * Verify an agent's identity and fetch their current trust score.
   * Returns a not-verified result for unknown agents (404) rather than throwing.
   */
  async verify(params: VerifyParams): Promise<Record<string, unknown>> {
    const headers = this.headers();
    if (params.signature) headers["X-Repuwave-Signature"] = params.signature;
    if (params.timestamp !== undefined) {
      headers["X-Repuwave-Timestamp"] = params.timestamp.toString();
    }
    if (params.bodyHash) headers["X-Repuwave-Body-Hash"] = params.bodyHash;

    const response = await fetch(`${this.baseUrl}/verify/${params.uaid}/`, {
      method: "GET",
      headers,
    });

    if (response.status === 404) {
      return { verified: false, score: 0, trust_level: "UNKNOWN", uaid: params.uaid };
    }
    if (!response.ok) return this.parseError(response);
    return response.json() as Promise<Record<string, unknown>>;
  }

  /**
   * Report an interaction outcome for an agent. Must include the agent's
   * original request signature and signed timestamp as proof of interaction.
   */
  async report(params: ReportParams): Promise<Record<string, unknown>> {
    let bodyHash = params.bodyHash ?? "";
    if (params.body && !bodyHash) {
      bodyHash = createHash("sha256").update(params.body).digest("hex");
    }

    const response = await fetch(`${this.baseUrl}/report/`, {
      method: "POST",
      headers: this.headers({ "Content-Type": "application/json" }),
      body: JSON.stringify({
        uaid: params.uaid,
        event_type: params.eventType,
        weight: params.weight,
        signature: params.signature,
        agent_timestamp: params.agentTimestamp,
        body_hash: bodyHash,
        timestamp: new Date().toISOString(),
      }),
    });

    if (!response.ok) return this.parseError(response);
    return response.json() as Promise<Record<string, unknown>>;
  }
}
